vpr/src/backend/physical_netlist_writer.h

```diff
@@ -0,0 +1,246 @@
+#ifndef WRITE_PHYSICAL_NETLIST_H
+#define WRITE_PHYSICAL_NETLIST_H
+
+#include <iostream>
+#include <list>
+#include <map>
+#include <ostream>
+#include <set>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "DeviceResources.capnp.h"
+#include "PhysicalNetlist.capnp.h"
+#include "capnp/serialize.h"
+#include "capnp/serialize-packed.h"
+#include "fpga_interchange_arch_utils.h"
+#include "netlist_walker.h"
+#include "netlist_writer.h"
+#include "route_tree_type.h"
+
+using namespace DeviceResources;
+using namespace capnp;
+
+/** @brief Writes Physical netlist interchange file based on the netlist name by walking the netlist.
+ */
+void write_physical_netlist(t_vpr_setup& vpr_setup);
+
+namespace ic {
+
+struct StringEnumerator {
+    std::vector<std::string> strings;
+    std::unordered_map<std::string, size_t> string_to_index;
+
+    size_t get_index(const std::string& s) {
+        auto result = string_to_index.emplace(s, strings.size());
+        if (result.second) {
+            // This string was inserted, append.
+            strings.push_back(s);
+        }
+
+        return result.first->second;
+    }
+
+    std::string get_string(size_t idx) {
+        return strings[idx];
+    }
+};
+
+enum e_route_type {
+    SITE_PIN = 0,
+    SITE_PIP = 1,
+    BEL_PIN = 2,
+    PIP = 3,
+    INVALID = 4
+};
+
+struct t_route_segment {
+    size_t tile = OPEN;
+    size_t site = OPEN;
+    size_t pin = OPEN;
+    size_t bel = OPEN;
+
+    // If is PIP
+    size_t wire0 = OPEN;
+    size_t wire1 = OPEN;
+
+    e_route_type type = INVALID;
+
+    std::string to_string(StringEnumerator& se) {
+        std::string name;
+
+        switch (type) {
+            case SITE_PIN:
+                name = "Site pin: site = " + se.get_string(site) + " pin = " + se.get_string(pin);
+                break;
+            case SITE_PIP:
+                name = "Site pip: site = " + se.get_string(site) + " pin = " + se.get_string(pin) + " bel = " + se.get_string(bel);
+                break;
+            case BEL_PIN:
+                name = "Bel pin: site = " + se.get_string(site) + " pin = " + se.get_string(pin) + " bel = " + se.get_string(bel);
+                break;
+            case PIP:
+                name = "Pip: tile = " + se.get_string(tile) + " wire0 = " + se.get_string(wire0) + " wire1 = " + se.get_string(wire1);
+                break;
+            default:
+                VTR_ASSERT(false);
+        }
+
+        return name;
+    }
```

Comment:
This should be in [vpr/src/backend/physical_netlist_writer.cpp](https://github.com/antmicro/vtr-verilog-to-routing/pull/16/files#diff-4ce1abf70f69a7f30df441750f88c3431b58aa7033aa70787265cd556b27cfb7)

Reply:
This can stay in the struct definition I believe, as the `route_segment` is not a class.